File: backend/app/Models/Purchases/Bill.php
<?php

namespace App\Models\Purchases;

use App\Models\Accounting\JournalEntry;
use App\Models\Company;
use App\Models\Currency;
use App\Models\User;
use App\Traits\BelongsToCompany;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\Relations\MorphMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class Bill extends Model
{
    use HasFactory, SoftDeletes, BelongsToCompany;

    protected $fillable = [
        'company_id',
        'vendor_id',
        'journal_entry_id',
        'bill_number',
        'vendor_invoice_number',
        'date',
        'due_date',
        'reference',
        'notes',
        'currency_id',
        'exchange_rate',
        'subtotal',
        'tax_amount',
        'discount_amount',
        'total',
        'paid_amount',
        'balance',
        'status',
        'approved_at',
        'approved_by',
        'paid_at',
    ];

    protected $casts = [
        'date' => 'date',
        'due_date' => 'date',
        'exchange_rate' => 'decimal:10',
        'subtotal' => 'decimal:4',
        'tax_amount' => 'decimal:4',
        'discount_amount' => 'decimal:4',
        'total' => 'decimal:4',
        'paid_amount' => 'decimal:4',
        'balance' => 'decimal:4',
        'approved_at' => 'datetime',
        'paid_at' => 'datetime',
    ];

    protected $attributes = [
        'exchange_rate' => 1,
        'subtotal' => 0,
        'tax_amount' => 0,
        'discount_amount' => 0,
        'total' => 0,
        'paid_amount' => 0,
        'balance' => 0,
        'status' => 'draft',
    ];

    /**
     * Status constants
     */
    const STATUS_DRAFT = 'draft';
    const STATUS_APPROVED = 'approved';
    const STATUS_PARTIAL = 'partial';
    const STATUS_PAID = 'paid';
    const STATUS_OVERDUE = 'overdue';
    const STATUS_CANCELLED = 'cancelled';

    /**
     * Relationships
     */
    public function company(): BelongsTo
    {
        return $this->belongsTo(Company::class);
    }

    public function vendor(): BelongsTo
    {
        return $this->belongsTo(Vendor::class);
    }

    public function currency(): BelongsTo
    {
        return $this->belongsTo(Currency::class);
    }

    public function journalEntry(): BelongsTo
    {
        return $this->belongsTo(JournalEntry::class);
    }

    public function approvedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'approved_by');
    }

    public function lines(): HasMany
    {
        return $this->hasMany(BillLine::class);
    }

    public function payments(): HasMany
    {
        return $this->hasMany(BillPayment::class);
    }

    public function attachments(): MorphMany
    {
        return $this->morphMany(\App\Models\Attachment::class, 'attachable');
    }

    /**
     * Calculate totals from lines
     */
    public function calculateTotals(): void
    {
        $this->subtotal = $this->lines->sum(fn($line) => $line->quantity * $line->unit_price - $line->discount_amount);
        $this->tax_amount = $this->lines->sum('tax_amount');
        $this->total = $this->subtotal + $this->tax_amount;
        $this->balance = $this->total - $this->paid_amount;
    }

    /**
     * Record payment
     */
    public function recordPayment(float $amount): void
    {
        $this->paid_amount += $amount;
        $this->balance = $this->total - $this->paid_amount;

        if ($this->balance <= 0.01) {
            $this->status = self::STATUS_PAID;
            $this->paid_at = now();
        } elseif ($this->paid_amount > 0) {
            $this->status = self::STATUS_PARTIAL;
        }

        $this->save();
    }

    /**
     * Approve bill
     */
    public function approve(User $user): void
    {
        if ($this->status === self::STATUS_DRAFT) {
            $this->status = self::STATUS_APPROVED;
            $this->approved_by = $user->id;
            $this->approved_at = now();
            $this->save();
        }
    }

    /**
     * Check if overdue
     */
    public function isOverdue(): bool
    {
        return $this->due_date < now() &&
               in_array($this->status, [self::STATUS_APPROVED, self::STATUS_PARTIAL]);
    }

    /**
     * Scopes
     */
    public function scopeUnpaid($query)
    {
        return $query->whereIn('status', [self::STATUS_APPROVED, self::STATUS_PARTIAL, self::STATUS_OVERDUE]);
    }

    public function scopeOverdue($query)
    {
        return $query->where('due_date', '<', now())
            ->whereIn('status', [self::STATUS_APPROVED, self::STATUS_PARTIAL]);
    }

    /**
     * Boot method
     */
    protected static function boot()
    {
        parent::boot();

        static::creating(function ($bill) {
            if (!$bill->bill_number) {
                $bill->bill_number = static::generateBillNumber($bill->company_id);
            }

            if (!$bill->due_date && $bill->vendor) {
                $bill->due_date = $bill->date->addDays($bill->vendor->payment_terms_days);
            }
        });

        static::saved(function ($bill) {
            if ($bill->isOverdue() && $bill->status !== self::STATUS_OVERDUE) {
                $bill->status = self::STATUS_OVERDUE;
                $bill->saveQuietly();
            }
        });
    }

    /**
     * Generate bill number
     */
    private static function generateBillNumber(int $companyId): string
    {
        $year = date('Y');
        $prefix = "BILL-{$year}-";

        $lastBill = static::where('company_id', $companyId)
            ->where('bill_number', 'like', $prefix . '%')
            ->orderByDesc('bill_number')
            ->first();

        if ($lastBill) {
            $lastNumber = (int) str_replace($prefix, '', $lastBill->bill_number);
            $newNumber = $lastNumber + 1;
        } else {
            $newNumber = 1;
        }

        return $prefix . str_pad($newNumber, 6, '0', STR_PAD_LEFT);
    }
}

File: desktop/lib/core/models/bill.dart
enum BillStatus {
  draft,
  pending,
  partial,
  paid,
  overdue,
  cancelled,
}

class Bill {
  final String id;
  final String billNumber;
  final String vendorId;
  final String? vendorName;
  final DateTime date;
  final DateTime dueDate;
  final double subtotal;
  final double taxAmount;
  final double total;
  final double amountPaid;
  final BillStatus status;
  final String currencyCode;
  final String? category;
  final List<BillLine> lines;
  final String? notes;
  final String? reference;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int? syncSequence;

  Bill({
    required this.id,
    required this.billNumber,
    required this.vendorId,
    this.vendorName,
    required this.date,
    required this.dueDate,
    this.subtotal = 0.0,
    this.taxAmount = 0.0,
    this.total = 0.0,
    this.amountPaid = 0.0,
    this.status = BillStatus.draft,
    this.currencyCode = 'UGX',
    this.category,
    required this.lines,
    this.notes,
    this.reference,
    required this.createdAt,
    required this.updatedAt,
    this.syncSequence,
  });

  double get balance => total - amountPaid;
  bool get isPaid => status == BillStatus.paid;
  bool get isOverdue => DateTime.now().isAfter(dueDate) && balance > 0;

  factory Bill.fromJson(Map<String, dynamic> json) {
    return Bill(
      id: (json['id'] ?? '').toString(),
      billNumber: json['bill_number'] as String,
      vendorId: (json['vendor_id'] ?? '').toString(),
      vendorName: json['vendor_name'] as String?,
      date: DateTime.parse(json['date'] as String),
      dueDate: DateTime.parse(json['due_date'] as String),
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0.0,
      taxAmount: (json['tax_amount'] as num?)?.toDouble() ?? 0.0,
      total: (json['total'] as num?)?.toDouble() ?? 0.0,
      amountPaid: (json['amount_paid'] as num?)?.toDouble() ?? 0.0,
      status: BillStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => BillStatus.draft,
      ),
      currencyCode: json['currency_code'] as String? ?? 'UGX',
      category: json['category'] as String?,
      lines: (json['lines'] as List<dynamic>?)
              ?.map((l) => BillLine.fromJson(l as Map<String, dynamic>))
              .toList() ??
          [],
      notes: json['notes'] as String?,
      reference: json['reference'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
      updatedAt: DateTime.parse(json['updated_at'] as String),
      syncSequence: json['sync_sequence'] as int?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bill_number': billNumber,
      'vendor_id': vendorId,
      'vendor_name': vendorName,
      'date': date.toIso8601String(),
      'due_date': dueDate.toIso8601String(),
      'subtotal': subtotal,
      'tax_amount': taxAmount,
      'total': total,
      'amount_paid': amountPaid,
      'status': status.name,
      'currency_code': currencyCode,
      'category': category,
      'lines': lines.map((l) => l.toJson()).toList(),
      'notes': notes,
      'reference': reference,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
      'sync_sequence': syncSequence,
    };
  }

  Bill copyWith({
    String? id,
    String? billNumber,
    String? vendorId,
    String? vendorName,
    DateTime? date,
    DateTime? dueDate,
    double? subtotal,
    double? taxAmount,
    double? total,
    double? amountPaid,
    BillStatus? status,
    String? currencyCode,
    String? category,
    List<BillLine>? lines,
    String? notes,
    String? reference,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? syncSequence,
  }) {
    return Bill(
      id: id ?? this.id,
      billNumber: billNumber ?? this.billNumber,
      vendorId: vendorId ?? this.vendorId,
      vendorName: vendorName ?? this.vendorName,
      date: date ?? this.date,
      dueDate: dueDate ?? this.dueDate,
      subtotal: subtotal ?? this.subtotal,
      taxAmount: taxAmount ?? this.taxAmount,
      total: total ?? this.total,
      amountPaid: amountPaid ?? this.amountPaid,
      status: status ?? this.status,
      currencyCode: currencyCode ?? this.currencyCode,
      category: category ?? this.category,
      lines: lines ?? this.lines,
      notes: notes ?? this.notes,
      reference: reference ?? this.reference,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      syncSequence: syncSequence ?? this.syncSequence,
    );
  }
}

class BillLine {
  final String id;
  final String billId;
  final String accountId;
  final String? accountName;
  final String description;
  final double quantity;
  final double unitPrice;
  final double amount;
  final double taxRate;
  final double taxAmount;

  BillLine({
    required this.id,
    required this.billId,
    required this.accountId,
    this.accountName,
    required this.description,
    this.quantity = 1.0,
    this.unitPrice = 0.0,
    this.amount = 0.0,
    this.taxRate = 0.0,
    this.taxAmount = 0.0,
  });

  factory BillLine.fromJson(Map<String, dynamic> json) {
    final account = json['account'] as Map<String, dynamic>?;
    final amount = (json['amount'] as num?)?.toDouble() ?? 0.0;
    return BillLine(
      id: (json['id'] ?? '').toString(),
      billId: (json['bill_id'] ?? '').toString(),
      accountId: (json['account_id'] ?? '').toString(),
      accountName: json['account_name'] as String? ?? account?['name'] as String?,
      description: json['description'] as String? ?? '',
      quantity: (json['quantity'] as num?)?.toDouble() ?? 1.0,
      unitPrice: (json['unit_price'] as num?)?.toDouble() ?? amount,
      amount: amount,
      taxRate: (json['tax_rate'] as num?)?.toDouble() ?? 0.0,
      taxAmount: (json['tax_amount'] as num?)?.toDouble() ?? 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'bill_id': billId,
      'account_id': accountId,
      'account_name': accountName,
      'description': description,
      'quantity': quantity,
      'unit_price': unitPrice,
      'amount': amount,
      'tax_rate': taxRate,
      'tax_amount': taxAmount,
    };
  }
}
